File: LAADS/VNP02IMG.A2020073.0942.002.2021126174430.nc.py
"""

This example code illustrates how to access and visualize a LAADS VNP02IMG 
netCDF-4/HDF5 Swath file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python VNP02IMG.A2020073.0942.002.2021126174430.nc.py

The netCDF-4/HDF5 data file and geolocation file [1] must be in your
current working directory.

Tested under: Python 3.9.13 :: Miniconda
Last updated: 2024-08-08
"""

import h5py
import os

import cartopy.crs as ccrs
import matplotlib.pyplot as plt
import numpy as np

from cartopy.mpl.gridliner import LATITUDE_FORMATTER, LONGITUDE_FORMATTER
from pyhdf.SD import SD, SDC

FILE_NAME = "VNP02IMG.A2020073.0942.002.2021126174430.nc"
GEO_FILE_NAME = "VNP03IMG.A2020073.0942.002.2021125004714.nc"
name = "/observation_data/I05"

with h5py.File(FILE_NAME, mode="r") as f:
    # Read dataset.
    data_raw = f[name][:]
    data = data_raw[:, :].astype(np.double)
    units = f[name].attrs['units']
    units = units.decode('ascii', 'replace')
    long_name = f[name].attrs['long_name']
    long_name = long_name.decode('ascii', 'replace')
    _FillValue = f[name].attrs['_FillValue']
    scale_factor = f[name].attrs['scale_factor']
    add_offset = f[name].attrs['add_offset']
    valid_max = f[name].attrs['valid_max']
    valid_min = f[name].attrs['valid_min']        
    invalid = np.logical_or(data > valid_max,
                            data < valid_min)
    invalid = np.logical_or(invalid, data == _FillValue)
    data[invalid] = np.nan
    data = np.ma.masked_where(np.isnan(data), data)
    data = scale_factor * data + add_offset
    
# Read lat/lon.
with h5py.File(GEO_FILE_NAME, mode="r") as f:
    # Get the geolocation data
    lat = f["/geolocation_data/latitude"][:]
    lon = f["/geolocation_data/longitude"][:]



# Find middle location.
lat_m = lat[int(lat.shape[0] / 2), int(lat.shape[1] / 2)]
lon_m = lon[int(lon.shape[0] / 2), int(lon.shape[1] / 2)]

# Let's use ortho projection.
orth = ccrs.Orthographic(
    central_longitude=lon_m, central_latitude=lat_m, globe=None
)
ax = plt.axes(projection=orth)

# Remove the following to see zoom-in view.
# ax.set_global()

# Plot on map.
p = plt.scatter(
    lon, lat, c=data, s=1, cmap=plt.cm.jet, transform=ccrs.PlateCarree()
)
# Put grids.
gl = ax.gridlines()

# Put coast lines.
ax.coastlines()


# Put grid labels only at left and bottom.
gl.top_labels = False
gl.right_labels = False

# Put degree N/E label.
gl.xformatter = LONGITUDE_FORMATTER
gl.yformatter = LATITUDE_FORMATTER

# Adjust colorbar size and location using fraction and pad.
cb = plt.colorbar(p, fraction=0.022, pad=0.01)

cb.set_label(units, fontsize=8)

# Put title.
basename = os.path.basename(FILE_NAME)
plt.title("{0}\n{1}".format(basename, long_name), fontsize=8)
fig = plt.gcf()

# Save image.
pngfile = "{0}.py.png".format(basename)
fig.savefig(pngfile)

# References
# [1] https://cmr.earthdata.nasa.gov/search/concepts/C2105092163-LAADS.html

File: LAADS/VNP02IMG_A2020073_0942_002_2021126174430_nc.m
%
% This example code illustrates how to access and visualize LAADS
% VNP02IMG v2 NetCDF-4/HDF5 Swath file in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example, please 
% use the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%  If you would like to see an  example of any other NASA HDF/HDF-EOS data 
% product that is not listed in the HDF-EOS Comprehensive Examples page 
% (http://hdfeos.org/zoo), feel free to contact us at eoshelp@hdfgroup.org or 
% post it at the HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage: save this script and run (without .m at the end)
% 
% $matlab -nosplash -nodesktop -r VNP02IMG_A2020073_0942_002_2021126174430_nc
%
% Tested under: MATLAB R2023b
% Last updated: 2024-08-09

% Open file.
FILE_NAME='VNP02IMG.A2020073.0942.002.2021126174430.nc';
ncid = netcdf.open(FILE_NAME, 'nowrite');

% Read data.
DATAFIELD_NAME='/observation_data/I05';
data = ncread(FILE_NAME, DATAFIELD_NAME);

% ncread() applies scale and offset automatically.
% Verify it using min/max.
% a = min(min(data))
% b = max(max(data))

% Read units.
units = ncreadatt(FILE_NAME, DATAFIELD_NAME, "units");

% Read long_name.
long_name = ncreadatt(FILE_NAME, DATAFIELD_NAME, "long_name");

% Read _FillValue.
fill_value = ncreadatt(FILE_NAME, DATAFIELD_NAME, "_FillValue");

% Read add_offset.
add_offset = ncreadatt(FILE_NAME, DATAFIELD_NAME, "add_offset");

% Read scale_factor.
scale_factor = ncreadatt(FILE_NAME, DATAFIELD_NAME, "scale_factor");

% Read valid_min.
valid_min = ncreadatt(FILE_NAME, DATAFIELD_NAME, "valid_min");

% Read valid_max.
valid_max = ncreadatt(FILE_NAME, DATAFIELD_NAME, "valid_max");

netcdf.close(ncid);

% Open geo-location file from [1].
GEO_FILE_NAME = 'VNP03IMG.A2020073.0942.002.2021125004714.nc';
ncid = netcdf.open(GEO_FILE_NAME, 'nowrite');

% Open the dataset.
Lat_NAME='geolocation_data/latitude';
Lon_NAME='geolocation_data/longitude';
lat=ncread(GEO_FILE_NAME, Lat_NAME);
lon=ncread(GEO_FILE_NAME, Lon_NAME);
% Close the file.
netcdf.close(ncid);

% Replace the fill value with NaN
dataf = double(data);
dataf(data == (double(fill_value) * scale_factor + add_offset)) = NaN;

% Handle valid range.
dataf(data < (double(valid_min) * scale_factor + add_offset)) = NaN;
dataf(data > (double(valid_max) * scale_factor + add_offset)) = NaN;

% Verify filtered values using min/max.
% a = min(min(dataf))
% b = max(max(dataf))


% Set the map parameters.
[xdimsize, ydimsize] = size(data);
lon_c = lon(xdimsize/2, ydimsize/2);
lat_c = lat(xdimsize/2, ydimsize/2);
latlim=ceil(max(max(lat))) - floor(min(min(lat)));

% Create the graphics figure -- 'visible'->'off' = off-screen rendering.
f=figure('Name', FILE_NAME, ...
	 'Renderer', 'zbuffer', ...
	 'Position', [0,0,800,600], ...
	 'visible', 'off');

% Use colormap.
colormap('Jet');

% Use FlatLimit for zoom-in effect in Ortho projection.
axesm ('ortho', 'Frame', 'on', 'Grid', 'on', ...
       'FLatLimit', [-Inf, latlim], ...       
       'origin', [lat_c, lon_c])
mlabel('equator')
plabel(0); 
plabel('fontweight','bold')

% Plot data. 
lats = lat(:);
lons = lon(:);
datas = dataf(:);

% Subset points to avoid system crashing due to memory.
% scatterm(lat, lon, 1, data);
% step = 100;
% step = 10;
step = 5;
% step = 2; killed

%scatterm(lat(1:step:end), lon(1:step:end), 1.0, dataf(1:step:end));
scatterm(lats(1:step:end), lons(1:step:end), 1.0, datas(1:step:end));


% Load the coastlines data file.
coast = load('coastlines.mat');

% Plot coastlines in color black ('k').
plotm(coast.coastlat, coast.coastlon, 'k')

h=colorbar();
set (get(h, 'title'), 'string', units);

% Set the title using long_name.
title({FILE_NAME; long_name}, ...
      'interpreter', 'none', 'FontSize',10,'FontWeight','bold');
saveas(f, [FILE_NAME '.m.png']);
exit;
